new messaging event api

snopoke opened the pull request:
## Summary
This replaces the current `messaging-events` API with a new one.

The old API was based on the `MessagingEvent` model which has some limitiations:
* When the recipient of the event is a Group, Case Group or Location there is only one event present and no data is available for individual members of the group / location.
* The data does not include the content of the actual messages or their status

[Spec doc](https://docs.google.com/document/d/1QWoLjfpl4_n4b_UcNpqZNfdZb0-2UyUMs5rgFb42kDY/edit#)

## Product Description
Updates the Messaging Event API to more closely match the Messaging History and detail reports.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
Added

### QA Plan
I will QA this on staging.

### Safety story
####Can we replace the existing API?
The existing API was added recently and has not had any usage on production yet.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

snopoke committed:
initial API structure

snopoke committed:
update content type and status

snopoke committed:
populate 'source' field

snopoke committed:
populate 'recipient' field

snopoke committed:
populate messages

snopoke committed:
remove app_id and additional_error_text

snopoke committed:
fix pop messages

snopoke committed:
test case rule sms

snopoke committed:
test survey event

snopoke committed:
test email

snopoke committed:
add todos

snopoke committed:
date filtering

snopoke committed:
source filtering

snopoke committed:
content type filter

snopoke committed:
status filtering

snopoke committed:
error_code filtering

snopoke committed:
test case_id filtering

snopoke committed:
cleanup

snopoke committed:
replace old API with new one

snopoke committed:
prefetch parent

snopoke committed:
rename form_unique_id to form_definition_id

snopoke committed:
lint

snopoke committed:
only support json output

czue commented:
I'd feel better if someone else reviewed, but I took a glance and this looks great! Quite an upgrdae from the previous version.

orangejenny commented on corehq/apps/api/tests/test_messaging_event_api.py:
```diff
@@ -80,6 +81,61 @@ def test_domain_filter(self):
+            "date.gt": dates[2].isoformat(),
+        }, [d.isoformat() for d in dates[3:]])
+
+    def test_date_filter_gte(self):
```
I don't quite follow this test. Won't all 5 test messages be created on today's date?

snopoke replied:
Yes they will. I added this test for symmetry with `test_date_filter_lte`. When supplying a date without the timestamp the API should include all messages on that date.

orangejenny commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -1254,6 +1258,33 @@ def _date_consumer(key, value):
+        if key != "source":
+            return
+
+        if ',' in value:
```
Isn't this whole if/else equivalent to `values = value.split(',')`?

snopoke replied:
5eff815550a

orangejenny commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -1089,26 +1095,296 @@ def prepend_urls(self):
 
 
 class MessagingEventResource(HqBaseResource, ModelResource):
```
Could this get a docstring clarifying that it returns messaging subevents?

snopoke replied:
1251af68a6c

snopoke committed:
add contact filter

snopoke commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -1160,19 +1160,71 @@ def dehydrate_error(self, bundle):
-        #         "contact": "", phone number / email
-        #     }
-        # ]
+        event = bundle.obj
```
None of the other content types have any sub-event data: CONTENT_ADHOC_SMS, CONTENT_API_SMS, CONTENT_CHAT_SMS, CONTENT_NONE, CONTENT_PHONE_VERIFICATION.

Those types only appear to be used on the MessagingEvent model.

snopoke committed:
simplify split

snopoke committed:
add message error_message

snopoke committed:
include 'partially_completed' status

snopoke commented on corehq/apps/smsforms/models.py:
```diff
@@ -163,6 +163,15 @@ def status(self):
+    @property
+    def status_api(self):
+        if self.submission_id:
+            return "complete"
```
5de52415132

snopoke committed:
display -> name

snopoke committed:
docstrings

snopoke commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -1089,26 +1096,357 @@ def prepend_urls(self):
+        else:
+            return {"contact": Q(sms__phone_number__contains=value)}
+
+        raise InvalidFilterError("Contact filter value must be a valid email address of phone number.")
```
```suggestion
        raise InvalidFilterError("Contact filter value must be a valid email address or phone number.")
```

snopoke committed:
fix typo

snopoke committed:
fix test assertion

calellowitz reviewed (commented):
Looks good, a few questions and a few nits but generally seems great

snopoke committed:
refactor xform session status to reduce duplication and use constants

snopoke commented on corehq/apps/smsforms/models.py:
```diff
@@ -163,6 +163,15 @@ def status(self):
+        elif self.session_is_open and self.session_type == XFORMS_SESSION_SMS:
+            return "in-progress"
+        else:
+            return "not-finished"
```
I'm not entirely sure. This is just duplicating what the `status` property does. I've refactored to remove the duplicate logic and to use constants: 649ac383e09

snopoke replied:
Following up again re `not-finished` vs `in-progress`, it appears this only applies to IVR events. It's been like this since it was added: https://github.com/dimagi/commcare-hq/commit/b5dae782fedcc9509421631669cf7f1d958e4741#diff-cd82d34a4d32a57355cf795d95d2a964b44b1919ac118e093c168d36209db36aR35

I'll ask Giovanni if he recalls: https://github.com/dimagi/commcare-hq/commit/b5dae782fedcc9509421631669cf7f1d958e4741#r51919687

snopoke committed:
use get instead of filter

snopoke commented on corehq/apps/reports/standard/message_event_display.py:
```diff
@@ -100,115 +101,136 @@ def get_sms_status_display(sms):
         EditNormalKeywordView,
     )
+    try:
+        keyword = Keyword.objects.filter(couch_id=keyword_id)
```
:+1: 03d58cfef0a

snopoke commented on corehq/apps/sms/models.py:
```diff
@@ -1516,6 +1516,12 @@ class MessagingSubEvent(models.Model, MessagingStatusMixin):
+    RECIPIENT_SLUGS = {
+        MessagingEvent.RECIPIENT_CASE: 'case',
+        MessagingEvent.RECIPIENT_MOBILE_WORKER: 'mobile-worker',
+        MessagingEvent.RECIPIENT_WEB_USER: 'web-user',
```
yes, they can receive emails.

snopoke committed:
rename test utils

snopoke commented on corehq/apps/reports/standard/message_event_display.py:
```diff
@@ -236,3 +258,45 @@ def get_event_display(domain, event, content_cache):
+        return display or "deleted-broadcast"
+    elif source == MessagingEvent.SOURCE_CASE_RULE:
+        display, _ = _get_case_rule_display_raw(domain, source_id)
+        return display or "deleted-conditional-alert"
```
these aren't referenced anywhere else in code so I don't really see the need to make them constants.

snopoke committed:
trim whitespace from contact filter value

snopoke committed:
split contact filter into phone_number and email

snopoke commented on corehq/apps/reports/tests/test_message_log_report.py:
```diff
@@ -79,64 +80,17 @@ def get_report_rows(self):
-            error=bool(error_message),
-            system_error_message=error_message,
-        )
+        sms = make_simple_sms(self.domain, message, error_message)
```
ff9afd83653

snopoke committed:
move messaging-event to standalone Django view

snopoke committed:
implement cursor based pagination

snopoke commented:
@orangejenny @calellowitz apologies for this next bit. I decided there wasn't much benefit to having this as a Tastypie view since I was doing all the filtering and serialization manually anyway so I've PR'd a change into this to replace the Tastypie view with a vanilla Django view: https://github.com/dimagi/commcare-hq/pull/29885. There is very little new code as most of it is just moving code around.

snopoke committed:
mark utils as 'nottest'

snopoke committed:
lint

snopoke committed:
lint

snopoke committed:
add missing return

snopoke committed:
Merge pull request #29886 from dimagi/sk/messaging-event-api-no-tastypie-cursor

implement cursor based pagination

snopoke committed:
Merge pull request #29885 from dimagi/sk/messaging-event-api-no-tastypie

convert messaging-event to standalone Django view

snopoke committed:
translate errors

snopoke committed:
convert namedtuple to attrs

snopoke committed:
add test for pagination with multiple matching dates at page break

snopoke committed:
organize imports

snopoke committed:
allow status filtering on all status values

snopoke committed:
make cursor URL absolute

snopoke committed:
auth headers instead of url params

snopoke committed:
serialize phone number and email address to separate fields

snopoke committed:
propogate message errors to main status

snopoke committed:
fix InternalTestMixin

snopoke committed:
add URL for detail view

orangejenny commented on corehq/apps/api/tests/internal_resources.py:
```diff
@@ -7,8 +7,8 @@
         # api auth should succeed
-        api_url = self._api_url(url, self.username)
-        response = self.client.get(api_url)
+        headers = self._get_api_key_auth_headers()
```
What does this change do?

snopoke replied:
this uses the `Authorization` header instead of passing the username and api key in URL params.

calellowitz reviewed (approved):
Looks good. Just to clarify my understanding, the cursor is passed from the server for the user to include in their next request as is. They do not have to modify the returned value at all?

Also, default ordering is date and id so it shouldn't hit any of the loops the DET does, but users can specify to just use date, and then it would be a risk?

snopoke commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -1408,6 +1408,9 @@ def _contact_filter_consumer(self, key, value):
             return
 
+        # '+' in URL get's converted to a space unless properly encoded
+        value = value.strip()
```
If you attempt to use the phone number filter with a value like `+9717488548` and don't encode the `+` then on the server side the filter value is ` 9717488548` (leading space).

snopoke commented:
> Looks good. Just to clarify my understanding, the cursor is passed from the server for the user to include in their next request as is. They do not have to modify the returned value at all?

That's correct, they can just take the value of `next` as it is which is the full URL of the next page.

> Also, default ordering is date and id so it shouldn't hit any of the loops the DET does, but users can specify to just use date, and then it would be a risk?

The 'id' ordering is always added regardless of whether or not they specify the ordering: https://github.com/dimagi/commcare-hq/pull/29869/files#diff-93ab9f3a28f576dba369ee8915991d807a79ffd5f4166a9ba0b8937528f33115R41

Ordering by ID isn't to prevent the looping issue, it's to allow us to return unique results in each page. Since the pagination is based on the filter `date.lte=<last date>` the page will always include the last element of the previous page. Without the ID sorting we can't reliably remove that element from the next page.

The looping issue can still happen if there was a full page of data that had exactly the same `date` value. This isn't an issue for this API since the `date` values are always set to the exact time of the event execution so it is unlikely that even 2 events would have the same date value.

snopoke committed:
add comments to status slugs

snopoke commented on corehq/apps/smsforms/models.py:
```diff
@@ -35,6 +35,22 @@ class SQLXFormsSession(models.Model):
 
+    STATUS_PARTIALLY_COMPLETE = "partially_completed"
+    STATUS_COMPLETE = "completed"
+    # "In Progress" means the survey session is open and the survey can still
+    # be completed by the user
+    STATUS_IN_PROGRESS = "in-progress"
+    # "Not Finished" means the survey session closed without generating a form submission
+    # and the survey can no longer be completed
+    STATUS_NOT_FINISHED = "not-finished"
```
@calellowitz FYI Giovanni responded with the distinction between 'in-progress' and 'not-finished'. I've added comments here to explain.

snopoke committed:
Merge branch 'master' into sk/messaging-event-api

snopoke committed:
rename 'id' fields to avoid issues with DET

DET uses some edge case feature of jsonpath_rw that interferes
with 'id' fields.

snopoke committed:
add message_id field

snopoke committed:
change missing vals to None

snopoke committed:
Merge branch 'master' into sk/messaging-event-api

snopoke committed:
update tests
